src/operator/rnn-inl.h

```diff
@@ -438,385 +565,893 @@ class RNNOp : public Operator{
+                                            dimA,
+                                            strideA));
+      #if USE_CUDNN_LSTM_PROJ
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(hy_desc_,
+                                            dtype_,
+                                            3,
+                                            dimB,
+                                            strideB));
+      #else
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(hy_desc_,
+                                            dtype_,
+                                            3,
+                                            dimA,
+                                            strideA));
+      #endif
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(cy_desc_,
+                                            dtype_,
+                                            3,
+                                            dimA,
+                                            strideA));
+      #if USE_CUDNN_LSTM_PROJ
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(dhx_desc_,
+                                            dtype_,
+                                            3,
+                                            dimB,
+                                            strideB));
+      #else
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(dhx_desc_,
+                                            dtype_,
+                                            3,
+                                            dimA,
+                                            strideA));
+      #endif
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(dcx_desc_,
+                                            dtype_,
+                                            3,
+                                            dimA,
+                                            strideA));
+      #if USE_CUDNN_LSTM_PROJ
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(dhy_desc_,
+                                            dtype_,
+                                            3,
+                                            dimB,
+                                            strideB));
+      #else
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(dhy_desc_,
+                                            dtype_,
+                                            3,
+                                            dimA,
+                                            strideA));
+      #endif
+      CUDNN_CALL(cudnnSetTensorNdDescriptor(dcy_desc_,
+                                            dtype_,
+                                            3,
+                                            dimA,
+                                            strideA));
+
+      // Create Dropout descriptors
+      DType* dropout_states_ = NULL;
+      if (param_.p > 0) {
+        CUDNN_CALL(cudnnDropoutGetStatesSize(s->dnn_handle_, &dropout_byte_));
+        dropout_size_ = dropout_byte_ / sizeof(DType);
+        dropout_states_ = ctx.requested[rnn_enum::kTempSpace].get_space_typed<xpu, 1, DType>(
```

Comment:
No, this is not the way to do it. As mentioned in the previous review, please refer to https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h#L249

Reply:
If using ctx.requested[rnn_enum::kTempSpace].get_cudnn_dropout_desc
            (&dropout_desc_, s, 1.0f - param_.p, seed_); to replace  dropout_states_ = ctx.requested[rnn_enum::kTempSpace].get_space_typed<xpu, 1, DType>(
            mshadow::Shape1(dropout_size_), s).dptr_;

gpu case will failed
[INFO] Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=1932186853 to reproduce.
test_operator_gpu.test_unfuse ... /home/hli32/mxnet/stateful/mxnet_gpu/python/mxnet/rnn/rnn_cell.py:675: UserWarning: NTC layout detected. Consider using TNC for FusedRNNCell for faster speed
  warnings.warn("NTC layout detected. Consider using "
[INFO] Setting test np/mx/python random seeds, use MXNET_TEST_SEED=1335409229 to reproduce.
ERROR
terminate called after throwing an instance of 'dmlc::Error'
  what():  [13:20:08] src/storage/storage.cc:134: Cannot Free space to a device you have not allocated

How to set dropout_states_ by get_cudnn_dropout_desc or something else?

Reply:
fixed